pkg/operator/controller/ingress_status.go

```diff
@@ -0,0 +1,273 @@
+	}
+
+	return ic, deploy, infraConfig, svcList.Items, nil
+}
+
+// computeIngressStatusConditions computes the ingress controller's current state.
+func computeIngressStatusConditions(conditions []operatorv1.OperatorCondition, ic *operatorv1.IngressController,
+	deployment *appsv1.Deployment, infraConfig *configv1.Infrastructure, services []corev1.Service) []operatorv1.OperatorCondition {
+
+	failingCondition := &operatorv1.OperatorCondition{
+		Type:   operatorv1.OperatorStatusTypeFailing,
+		Status: operatorv1.ConditionUnknown,
+	}
+	if ic != nil {
+		failingCondition.Status = operatorv1.ConditionFalse
+		failingCondition.Reason = "IngressFound"
+		failingCondition.Message = fmt.Sprintf("IngressController %q exists.", ic.Name)
+	} else {
+		failingCondition.Status = operatorv1.ConditionTrue
+		failingCondition.Reason = "NoIngress"
+		failingCondition.Message = "IngressController does not exist."
+	}
+	conditions = setIngressStatusCondition(conditions, failingCondition)
+
+	progressingCondition := &operatorv1.OperatorCondition{
+		Type:   operatorv1.OperatorStatusTypeProgressing,
+		Status: operatorv1.ConditionUnknown,
+	}
+	current := deployment.Status.AvailableReplicas
+	desired := *deployment.Spec.Replicas
+	if current == desired {
+		progressingCondition.Status = operatorv1.ConditionFalse
+		progressingCondition.Reason = "DeploymentComplete"
+		progressingCondition.Message = "Deployment completed required replicas."
+	} else {
+		progressingCondition.Status = operatorv1.ConditionTrue
+		progressingCondition.Reason = "DeploymentInProgress"
+		progressingCondition.Message = fmt.Sprintf("have %d Deployments, want %d", current, desired)
+	}
+	conditions = setIngressStatusCondition(conditions, progressingCondition)
+
+	lbManagedCondition := &operatorv1.OperatorCondition{
+		Type:   operatorv1.LoadBalancerManagedIngressConditionType,
+		Status: operatorv1.ConditionUnknown,
+	}
+	// Private endpoint publishing does not contain a service.
+	if ic.Status.EndpointPublishingStrategy.Type == operatorv1.PrivateStrategyType {
+		lbManagedCondition.Status = operatorv1.ConditionFalse
+		lbManagedCondition.Reason = "LoadBalancerServiceNotRequired"
+		lbManagedCondition.Message = "LoadBalancer Service not required by publishing strategy"
+	} else {
+		lbManagedCondition.Status = operatorv1.ConditionTrue
+		lbManagedCondition.Reason = "LoadBalancerServiceRequired"
+		lbManagedCondition.Message = "LoadBalancer Service required by publishing strategy"
+	}
+
+	conditions = setIngressStatusCondition(conditions, lbManagedCondition)
+
+	lbReadyCondition := &operatorv1.OperatorCondition{
+		Type:   operatorv1.LoadBalancerReadyIngressConditionType,
+		Status: operatorv1.ConditionUnknown,
+	}
+	if len(services) == 0 {
+		lbReadyCondition.Status = operatorv1.ConditionFalse
+		lbReadyCondition.Reason = "ServiceMissing"
+		lbReadyCondition.Message = "LoadBalancer Service does not exist."
+	} else {
+		for _, s := range services {
```

Comment:
The relationship between an ingresscontroller and any desired LB service is 1:1, not 1:M

Reply:
My thinking was multiple services are related to an ingresscontroller depending on the endpoint publishing strategy. Therefore, multiple services may be passed to `computeIngressStatusConditions`. This would make it possible to create a condition based on the status of the internal (`ClusterIP`) service. Thoughts?

Reply:
The operator should only be created 0 or 1 LB service for each ingresscontroller.  If the administrator creates an LB service, I don't think we should worry about it.  So I'm, thinking `getIngressState` should be figuring out which LB service, if any, is associated with the ingresscontroller, and `computeIngressStatusConditions` should take exactly one service as input.

Reply:
Since the condition is named `LoadBalancerReady`, I don't think the internal service should be considered in determining its status.